### PR TITLE
Remove global variables in IpOpt interface - issue #175

### PR DESCRIPTION
**Pull request by cpizzolato**
Removes the global variables in IpLapackSolverInterface.cpp. Makes `ipiv` a member variable of `IpLapackSolverInterface`. IpOpt can now be used in multiple concurrent threads. I have tested this with OpenSim inverse kinematics and with the Simbody test suit without encountering any problem.

### CODE REVIEW

**Comment by chrisdembia** [comment]
Do you know if IpOpt has addressed this issue in newer versions of IpOpt?

**Comment by cpizzolato** [comment]
Hi Chris,

the version of IpOpt currently implemented in simbody is pretty old, from the file headers seems from 2006. Comparing it with the latest version (3.12.5) of IpOpt (http://www.coin-or.org/download/source/Ipopt/) is not really straightforward, since the structure is quite different. I've checked the changelog and they mention some fix regarding thread-safety, although on a different file. Also, the file that I've patched no longer exists in the laterst IpOpt implementation.

**Comment by chrisdembia** [comment]
Okay; thank you for looking into that.

File: SimTKmath/Optimizers/src/IpOpt/IpLapackSolverInterface.cpp [hunk 1]
```diff
@@ -146,6 +146,7 @@ double *afact;
     delete [] a;
     delete [] irn_;
     delete [] jcn_;
+    delete [] ipiv_;
```
Comment (sherm1) [inline SimTKmath/Optimizers/src/IpOpt/IpLapackSolverInterface.cpp]: Should set `ipiv_ = nullptr;` here -- the other three are reassigned but ipiv is not.

File: SimTKmath/Optimizers/src/IpOpt/IpLapackSolverInterface.hpp [hunk 2]
```diff
@@ -160,6 +160,7 @@ namespace Ipopt
     //MUMPS data structure
 //    DMUMPS_STRUC_C mumps_data;
 
+    int *ipiv_;
```
Comment (sherm1) [inline SimTKmath/Optimizers/src/IpOpt/IpLapackSolverInterface.hpp]: For consistency, please move this up next to the other data members (below `negevals_`), with a comment like `/** Array for storing the pivot order after factorization. */`.

**Comment by sherm1** [comment]
Claudio, thank you very much for this PR! I'm done reviewing it -- looks great pending two small changes. Please let me know when those are in so we can merge this.

Question: is this actually sufficient to provide thread safety when using IpOpt or are there other problems remaining?

Regarding this version of IpOpt vs. the current one, I think the broken file here might be one we added. I looked at "blame" and it appears to have been created by Jack Middleton who worked on Simbody a long time ago.

**Comment by cpizzolato** [comment]
Sherm, thanks for the review! I'm going to update the PR with the suggested changes. 

I don't know if this modification solves all the thread safety issues since I tested it only through OpenSim. I'll try to build a test-case using just simbody to be sure there are no other problems and I'll get back to you.

Claudio

**Comment by chrisdembia** [comment]
Note: it seems like any other fixes for thread safety could happen in separate PRs.

**Comment by sherm1** [comment]
Agreed -- there is no question that this PR is an improvement. I was just wondering whether this was the end of the thread un-safety story.

**Comment by sherm1** [comment]
I would be very happy to merge this PR after the few minor changes we discussed.

**Comment by cpizzolato** [comment]
I've created a small test based on the example ConstrainedOptimization.cpp where I run multiple identical threads that perform the optimisation at the same time. With the master branch I get a seg fault, as expected. The patch solves the problem and everything seems fine. 

I can adapt the example so it can be added to the ctest if you think it makes sense. If so, I'll create another PR for it.

**Comment by cpizzolato** [comment]
Hi @sherm1, I've added a simple adhoc test based on the Ipopt test in the regression suite. It's basically the same, but it launches 10 optimizations concurrently and then it collects the results. Using the unsafe Ipopt in the master branch you will get a segmentation fault. Using the fixed Ipopt you get the same optimization results 10 times, as expected. 

Let me know if there's anything else I can do or fix reguarding this PR!

Cheers
Claudio

**Comment by sherm1** [comment]
@cpizzolato, thanks very much for this fix! The test case looks great. Sorry for the reviewing delay.

**Comment by cpizzolato** [comment]
@sherm1 thanks for the merge! I'm glad the test case was up to standards!

**Pull request merged**